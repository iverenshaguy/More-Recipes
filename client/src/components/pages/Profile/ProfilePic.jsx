import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { userPropTypes } from '../../../helpers/proptypes';
import { uploadValidation } from '../../../helpers/validations';
import { fileEventAdapter as adaptFileEventToValue, imageUpload } from '../../../utils';

/**
 * @exports
 * @class ProfilePic
 * @extends Component
 * @classdesc Returns User ProfilePic
 * @returns {JSX} User ProfilePic
 */
class ProfilePic extends Component {
  static propTypes = {
    ...userPropTypes,
    uploadImage: PropTypes.shape({
      error: PropTypes.string,
      success: PropTypes.bool
    }).isRequired,
    updateUserImage: PropTypes.func.isRequired
  }

  /**
   * @memberof ProfilePic
   * @constructor
   * @returns {JSX} ProfilePic
   */
  constructor() {
    super();

    this.state = {
      uploadTask: {},
      uploading: false,
      uploadError: null,
    };

    this.altImage = '/images/user-image-placeholder.png';
    this.handleImageUpload = this.handleImageUpload.bind(this);
    this.handleChangeImage = this.handleChangeImage.bind(this);
    this.handleChangeImageClick = this.handleChangeImageClick.bind(this);
    this.handleCancelImageUpload = this.handleCancelImageUpload.bind(this);
  }

  /**
   * @memberof ProfilePic
   * @param {object} e
   * @returns {nothing} Returns nothing
   */
  handleChangeImageClick(e) {
    e.preventDefault();
    this.setState({
      uploadError: null
    });

    if (!this.state.uploading) {
      return this.imageUploader.click();
    }
    return this.handleCancelImageUpload();
  }

  /**
   * @memberof ProfilePic
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChangeImage(event) {
    const file = event.target.files[0];
    const maxSize = 2 * 1024 * 1024; // 2MB max size
    const allowedTypes = ['image/gif', 'image/jpeg', 'image/png'];

    if (uploadValidation.call(this, file, maxSize, allowedTypes)) {
      adaptFileEventToValue(this.handleImageUpload, this.userImage)(event);
    }
  }

  /**
   * @memberof ProfilePic
   * @param {object} image - image
   * @return {state} returns new state
   */
  handleImageUpload(image) {
    imageUpload.call(
      this,
      image,
      this.props.user.profilePic,
      `users/${Date.now()}`,
      (downloadURL) => {
        // upload imageurl to db
        this.props.updateUserImage(this.props.user.id, downloadURL)
          .then(() => {
            if (!this.props.uploadImage.uploading && this.props.uploadImage.success) {
              this.setState({ uploading: false });
            }

            if (!this.props.uploadImage.uploading && !this.props.uploadImage.success) {
              this.setState({ uploadError: 'Something happened, please try again', uploading: false });
              this.state.uploadTask.delete();
            }
          });
      }
    );
  }

  /**
   * @memberof ProfilePic
   * @return {state} returns new state
   */
  handleCancelImageUpload() {
    const { user } = this.props;
    this.state.uploadTask.cancel();
    // change image source back to former image
    this.userImage.src = user.profilePic ? user.profilePic : this.altImage;
  }

  /**
   * @memberof ProfilePic
   * @returns {JSX} User ProfilePic
   */
  render() {
    const { user } = this.props;
    const { uploading, uploadError } = this.state;

    return (
      <div className="col-5 col-md-4 profile-picture align-self-center">
        <img
          src={user.profilePic ? user.profilePic : this.altImage}
          ref={(ref) => { this.userImage = ref; }}
          style={{ opacity: uploading ? 0.3 : 1 }}
          className="d-block d-md-inline rounded-circle img-fluid img-thumbnail user-img align-middle"
          alt="profile"
        />
        &nbsp; &nbsp;
        {/* {uploading &&
        !uploadError &&
        <Progress color="light" value={uploadProgress} className="w-50" />} */}
        <div className="d-block d-md-inline-block align-middle">
          {!uploading && uploadError && <p className="text-danger upload-error">{uploadError}</p>}
          <a href="#profile-photo" onClick={this.handleChangeImageClick}>{!uploading ? 'Change Picture' : 'Cancel Upload'}</a>
        </div>
        <input
          type="file"
          accept="image/gif, image/jpeg, image/png"
          ref={(ref) => { this.imageUploader = ref; }}
          onChange={this.handleChangeImage}
          style={{ display: 'none' }}
        />
      </div>);
  }
}

export default ProfilePic;
