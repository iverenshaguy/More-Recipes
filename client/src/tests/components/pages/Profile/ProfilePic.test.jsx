import React from 'react';
import ProfilePic from '../../../../components/pages/Profile/ProfilePic';

const mockFn = jest.fn();
const setup = () => {
  const props = {
    user: {
      id: 2,
      username: 'username'
    },
    uploadImageObj: {
      uploading: false,
      success: false,
      error: null
    },
    uploadImage: jest.fn(),
    setUploading: jest.fn(),
    uploadFailure: jest.fn(),
    uploadSuccess: jest.fn(),
    unsetUploading: jest.fn(),
    updateUserImage: mockFn,
    clearUploadError: jest.fn(),
  };

  const shallowWrapper = shallow(<ProfilePic {...props} />);
  const mountedWrapper = mount(<ProfilePic {...props} />);

  return { props, shallowWrapper, mountedWrapper };
};

const state = {
  uploadTask: {
    delete: jest.fn(),
    cancel: jest.fn()
  },
  uploading: false,
  uploadError: null
};

describe('ProfilePic', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('renders error when uploading is false and there is an error', () => {
    const { shallowWrapper } = setup();

    shallowWrapper.setState({ uploading: false, uploadError: 'An error occurred' });

    expect(shallowWrapper.find('.upload-error').length).toBeTruthy();
    expect(shallowWrapper.find('.upload-error').text()).toEqual('An error occurred');
  });

  it('does not render preloader when not uploading', () => {
    const { shallowWrapper } = setup();

    shallowWrapper.setProps({ uploadImageObj: { uploading: false } });

    expect(shallowWrapper.find('a[href="#profile-photo"]').text()).not.toEqual('<MiniPreLoader />');
    expect(shallowWrapper.find('a[href="#profile-photo"]').text()).toEqual('<FontAwesome />');
  });

  it('renders preloader when uploading', () => {
    const { shallowWrapper } = setup();

    shallowWrapper.setProps({ uploadImageObj: { uploading: true } });

    expect(shallowWrapper.find('a[href="#profile-photo"]').text()).toEqual('<MiniPreLoader />');
    expect(shallowWrapper.find('a[href="#profile-photo"]').text()).not.toEqual('<FontAwesome />');
  });

  it('renders user\'s profilepic when provided', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<ProfilePic
      {...props}
      user={{ id: 1, username: 'username', profilePic: 'pic.png' }}
    />, rrcMock.get());

    shallowWrapper.setState({ uploading: true });

    expect(shallowWrapper.find('img[src="pic.png"]')).toBeTruthy();
  });

  it('handles change image click', () => {
    const { mountedWrapper } = setup();
    const handleChangeImageClickSpy = jest.spyOn(mountedWrapper.instance(), 'handleChangeImageClick');
    const mockClick = jest.fn();

    mountedWrapper.instance().imageUploader.click = mockClick;
    mountedWrapper.setState({ ...state, uploadError: 'Error' });
    mountedWrapper.find('a[href="#profile-photo"]').simulate('click');

    expect(mountedWrapper.instance().state.uploadError).toBeNull();
    expect(handleChangeImageClickSpy).toHaveBeenCalled();
    expect(mockClick).toHaveBeenCalled();
    mountedWrapper.unmount();
  });

  it('handles change image', () => {
    const { mountedWrapper } = setup();
    const handleChangeImageSpy = jest.spyOn(mountedWrapper.instance(), 'handleChangeImage');
    const image = new Blob(['../../assets/images/user-image-placeholder.png'], { type: 'application/png' });
    const event = { target: { files: [image] } };

    mountedWrapper.setState(state);
    mountedWrapper.find('input[type="file"]').simulate('change', event);

    expect(handleChangeImageSpy).toHaveBeenCalled();
    mountedWrapper.unmount();
  });
});
