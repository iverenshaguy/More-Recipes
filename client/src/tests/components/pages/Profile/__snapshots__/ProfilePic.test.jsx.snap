// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePic renders correctly 1`] = `
<div
  className="col-5 col-md-4 profile-picture align-self-center"
>
  <img
    alt="profile"
    className="d-block d-md-inline rounded-circle img-fluid img-thumbnail user-img align-middle"
    src="/images/user-image-placeholder.png"
    style={
      Object {
        "opacity": 1,
      }
    }
  />
     
  <div
    className="d-block d-md-inline-block align-middle"
  >
    <a
      href="#profile-photo"
      onClick={[Function]}
    >
      Change Picture
    </a>
  </div>
  <input
    accept="image/gif, image/jpeg, image/png"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</div>
`;
